internal/scheduler/schedulercontexts.go

```diff
@@ -0,0 +1,364 @@
+	// Total number of jobs successfully scheduled in this round.
+	NumScheduledJobs int
+	// Reason for why the scheduling round finished.
+	TerminationReason string
+	// Protects everything in this struct.
+	mu sync.Mutex
+}
+
+func NewSchedulingContext(
+	executorId string,
+	totalResources schedulerobjects.ResourceList,
+	priorityFactorByQueue map[string]float64,
+	initialResourcesByQueueAndPriority map[string]schedulerobjects.QuantityByPriorityAndResourceType,
+) *SchedulingContext {
+	queueSchedulingContexts := make(map[string]*QueueSchedulingContext)
+	for queue := range priorityFactorByQueue {
+		queueSchedulingContexts[queue] = NewQueueSchedulingContext(
+			queue,
+			executorId,
+			priorityFactorByQueue[queue],
+			initialResourcesByQueueAndPriority[queue],
+		)
+	}
+	return &SchedulingContext{
+		Started:                      time.Now(),
+		ExecutorId:                   executorId,
+		QueueSchedulingContexts:      queueSchedulingContexts,
+		TotalResources:               totalResources.DeepCopy(),
+		ScheduledResourcesByPriority: make(schedulerobjects.QuantityByPriorityAndResourceType),
+	}
+}
+
+func (sctx *SchedulingContext) String() string {
+	var sb strings.Builder
+	w := tabwriter.NewWriter(&sb, 1, 1, 1, ' ', 0)
+	fmt.Fprintf(w, "Started:\t%s\n", sctx.Started)
+	fmt.Fprintf(w, "Finished:\t%s\n", sctx.Finished)
+	fmt.Fprintf(w, "Duration:\t%s\n", sctx.Finished.Sub(sctx.Started))
+	fmt.Fprintf(w, "Total capacity:\t%s\n", sctx.TotalResources.CompactString())
+	fmt.Fprintf(w, "Jobs scheduled:\t%d\n", sctx.NumScheduledJobs)
+	fmt.Fprintf(w, "Total resources scheduled:\t%s\n", sctx.ScheduledResourcesByPriority.AggregateByResource().CompactString())
+	fmt.Fprintf(w, "Total resources scheduled (by priority):\t%s\n", sctx.ScheduledResourcesByPriority.String())
+	fmt.Fprintf(
+		w, "Scheduled queues:\t%v\n",
+		maps.Keys(
+			armadamaps.Filter(
+				sctx.QueueSchedulingContexts,
+				func(_ string, qctx *QueueSchedulingContext) bool {
+					return len(qctx.SuccessfulJobSchedulingContexts) > 0
+				},
+			),
+		),
+	)
+	fmt.Fprintf(w, "Termination reason:\t%s\n", sctx.TerminationReason)
+	w.Flush()
+	return sb.String()
+}
+
+// AddJobSchedulingContext adds a job scheduling context.
+// Automatically updates scheduled resources
+func (sctx *SchedulingContext) AddJobSchedulingContext(jctx *JobSchedulingContext, isEvictedJob bool) {
+	sctx.mu.Lock()
+	defer sctx.mu.Unlock()
+	if !isEvictedJob && jctx.UnschedulableReason == "" {
+		sctx.ScheduledResourcesByPriority.AddResourceList(
+			jctx.Req.Priority,
+			schedulerobjects.ResourceListFromV1ResourceList(jctx.Req.ResourceRequirements.Requests),
+		)
+		sctx.NumScheduledJobs++
+	}
+	if qctx := sctx.QueueSchedulingContexts[jctx.Job.GetQueue()]; qctx != nil {
+		qctx.AddJobSchedulingContext(jctx, isEvictedJob)
+	}
+}
+
+// ClearJobSpecs zeroes out job specs to reduce memory usage.
+func (sctx *SchedulingContext) ClearJobSpecs() {
```

Comment:
To me this is a bit of a code smell.  I think the issue here is that the scheduling context provides two main functions.  A) to act as a store of state during the scheduler cycle and B) to provide a record of what we did.  Currently we're conflating these two functions into a single struct which means we have to do things like this.  I think ideally we have a separate set of object that represent reports and transform into those at the at the end of the scheduling cycle.  Obviously there would be a fair amount of overlap in the fields each would store, but it would means that we can have specialised structs for each which would avoid the need to do stuff like this and have other benefits (e.g. reports really should be immutable, while the context probably shouldnt).

  In the medium term I think we're going to need to do the above regardless, because  we probably want to expose the report struct over grpc and in this case we almost certainly don't want to expose our internal scheduling context.

Reply:
Agreed. Would prefer to do that in a follow-up PR.